src/workerd/api/streams/readable-source-adapter.c++

```diff
@@ -0,0 +1,1097 @@
+}
+
+kj::Maybe<uint64_t> ReadableStreamSourceKjAdapter::tryGetLength(StreamEncoding encoding) {
+  KJ_IF_SOME(active, state.tryGet<kj::Own<Active>>()) {
+    if (active->previousReadWasLast) {
+      // If the previous read indicated that it was the last, then
+      // let's just transition to the closed state now and return kj::none.
+      state.init<Closed>();
+      return kj::none;
+    }
+    KJ_IF_SOME(exc, active->pendingCancel) {
+      state = kj::mv(exc);
+      return kj::none;
+    }
+    return active->stream->tryGetLength(encoding);
+  }
+
+  // The stream is either closed or errored.
+  return kj::none;
+}
+
+void ReadableStreamSourceKjAdapter::cancel(kj::Exception reason) {
+  KJ_IF_SOME(active, state.tryGet<kj::Own<Active>>()) {
+    active->cancel(kj::cp(reason));
+  }
+  state = kj::mv(reason);
+}
+
+kj::Promise<void> ReadableStreamSourceKjAdapter::pumpToImpl(WritableStreamSink& output, bool end) {
+  static constexpr size_t kMinRead = 8192;
+  static constexpr size_t kMaxRead = 16384;
+  kj::FixedArray<kj::byte, kMaxRead> buffer;
+  // Let's make sure we're in the right state before we start.
+  KJ_DASSERT(state.is<kj::Own<Active>>());
+  bool writeFailed = false;
+  try {
+    while (true) {
+      // Check our state before each iteration of the loop. This
+      // is a bit redundant since the canceler should take care
+      // of aborting the loop if we are canceled, but it's good
+      // to be extra careful. If this proves to be a performance
+      // problem, we can wrap the KJ_SWITCH_ONEOF if an if KJ_DEBUG
+      // define so that the additional checks are only done in
+      // debug builds.
+      KJ_SWITCH_ONEOF(state) {
+        KJ_CASE_ONEOF(active, kj::Own<Active>) {
+          // Read at least 8192 bytes up to the size of our buffer.
+          // Keep in mind that tryRead() requires acquiring the isolate lock
+          // on each call, so we want to try to read a decent amount each time
+          // to avoid excessive lock latency. But, we also don't want to read
+          // too much and end up with too much memory pressure or lock latency.
+          // The values here are somewhat arbitrary, but seem reasonable.
+          auto bytesRead = co_await tryReadImpl(*active, buffer, kMinRead);
+
+          // WARNING: do not access 'active' after this point because the stream
+          // may have been closed or canceled while we were awaiting the read.
+          // Nothing below depends on active, so this is currently safe, but we
+          // will have to be careful. The next iteration of the loop will check
+          // the state again.
+
+          // Only write if we actually read something.
+          if (bytesRead > 0) {
+            try {
+              co_await output.write(buffer.asPtr().slice(0, bytesRead));
+            } catch (...) {
+              auto ex = kj::getCaughtExceptionAsKj();
+              writeFailed = true;
+              kj::throwFatalException(kj::cp(ex));
+            }
```

Comment:
Do we need to cancel Active on exceptions, like in the case of `output.end()` failing below?

If not, then this try-catch could just be
```cpp
KJ_ON_SCOPE_FAILURE(writeFailed = true);
co_await output.write(buffer.asPtr().slice(0, bytesRead));
```

It'd be nice to get rid of the `output.end()` try-catch, too, so we don't have nested try-catches so close to each other. But maybe that's not feasible, without a separate `endFailed` boolean.